core/app/models/spree/product.rb

```diff
@@ -323,7 +323,8 @@ def taxons_for_store(store)
     end
 
     def any_variant_in_stock_or_backorderable?
-      variants_including_master.in_stock_or_backorderable.exists?
+      only_master_variant = variants_including_master.count == 1 && variants.count.zero?
```

Comment:
This is prone to failure - please check `variants.any?` - this will use proper check for existence of any non master variants